### PR TITLE
Support DDP ignored parameters in DDPOptimizer

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88511
* #88480
* __->__ #88460
* #88435

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

File: torch/_dynamo/optimizations/distributed.py [hunk 1]
```diff
@@ -88,10 +100,16 @@ def compile_fn(self, gm: fx.GraphModule, example_inputs: List[torch.Tensor]):
             if node.op == "call_module":
                 target = gm.get_submodule(node.target)
                 for name, p in target.named_parameters():
+                    if (str(node), name) in self.parameters_to_ignore:
```
Comment (davidberard98) [inline torch/_dynamo/optimizations/distributed.py:103]: mirroring chat thread, @zdevito and @voznesenskym recommended `node.target in self.parameters_to_ignore` to avoid relying on fx naming behavior

File: torch/_dynamo/optimizations/distributed.py [hunk 2]
```diff
@@ -41,13 +42,20 @@ def pretty_print_buckets(buckets: List[Bucket]):
...
+
+
 class DDPOptimizer:
     def __init__(
         self,
         bucket_bytes_cap: int,
         parameters_to_ignore: List[str],
         backend_compile_fn,
-        debug=False,
+        debug=True,
```
Comment (aazzolini) [inline torch/_dynamo/optimizations/distributed.py:58]: Is this change intended?

File: torch/_dynamo/optimizations/distributed.py [hunk 3]
```diff
@@ -41,13 +42,20 @@ def pretty_print_buckets(buckets: List[Bucket]):
         print("Please `pip install tabulate` in order to pretty-print ddp bucket sizes")
 
 
+def fx_name(fqn):
```
Comment (aazzolini) [inline torch/_dynamo/optimizations/distributed.py:45]: This looks a bit fragile. Is this a standard thing in pytorch? Could you add a comment pointtin to the code that defines this format?

File: torch/_dynamo/optimizations/distributed.py [hunk 4]
```diff
@@ -41,13 +42,20 @@ def pretty_print_buckets(buckets: List[Bucket]):
...
+    parts = fqn.split(".")
+    module = "self_" + "_".join(parts[:-1])
```
Comment (aazzolini) [inline torch/_dynamo/optimizations/distributed.py:47]: Oh this is another one that is a bit fragile. Could you point to the place that defines this format?

File: torch/_dynamo/optimizations/distributed.py (hunk 3 above)
Comment (aazzolini) [reply torch/_dynamo/optimizations/distributed.py:45]: I guess let's make this explicitly private? _fx_name?

File: torch/_dynamo/optimizations/distributed.py (hunk 1 above)
Comment (aazzolini) [reply torch/_dynamo/optimizations/distributed.py:103]: oh i see . def we should do this then.

**Review by aazzolini — CHANGES_REQUESTED** [review]

File: torch/_dynamo/optimizations/distributed.py (hunk 2 above)
Comment (wconstab) [reply torch/_dynamo/optimizations/distributed.py:58]: oops, i thought i updated this back. thanks!

File: torch/_dynamo/optimizations/distributed.py (hunk 1 above)
Comment (wconstab) [reply torch/_dynamo/optimizations/distributed.py:103]: yea, i wish. node.target is `self_net_2` and name is `weight`
DDP tells us to ignore "net.2.weight".

I agree that this fx_name thing is unappealing.

File: test/distributed/test_dynamo_distributed.py [hunk 5]
```diff
@@ -386,6 +391,39 @@ def fn():
...
+        """
+        Verifies ddp graph-split logic ignores parameters marked to ignore on DDP module.
+        Hooks up graph-split optimizer manually so it can peek at internal state.
+        """
+        m, inputs, correct_outputs = get_custom_model(self.device)
+        parameters_to_ignore = ["seq.2.weight", "seq.4.linear.bias"]
+        DDP._set_params_and_buffers_to_ignore_for_model(m, parameters_to_ignore)
+        ddp_m = DDP(m, device_ids=self.device_ids, bucket_cap_mb=25)
+        parameter_ids_to_ignore = [
+            id(ddp_m.module.get_parameter(p))
```
Comment (wconstab) [inline test/distributed/test_dynamo_distributed.py:405]: so.. this seems better than the hacky fqn/mangled name thing.  But, is it totally reliable?  I wondered if there could be edge cases, or if dynamo would possibly make copies, etc.

cc @voznesenskym @zdevito

File: torch/_dynamo/eval_frame.py [hunk 6]
```diff
@@ -240,7 +240,10 @@ def catch_errors(frame, cache_size):
                 with compile_lock:
                     ddp_optimizer = DDPOptimizer(
                         bucket_bytes_cap=ddp_module.bucket_bytes_cap,
-                        parameters_to_ignore=ddp_module.parameters_to_ignore,
+                        parameter_ids_to_ignore=[
+                            id(ddp_module.module.get_parameter(p))
+                            for p in ddp_module.parameters_to_ignore
+                        ],
```
Comment (voznesenskym) [inline torch/_dynamo/eval_frame.py:246]: if these parameters are id stable, why not just annotate them directly? at the ddp level, whenever I add to `parameters_to_ignore` it should be something like:

```
def mark_parameter_as_ignored(module, name):
    assert name in module.named_parameters()
    ignored_parameter_list.append(name)
    parameter = module.named_parameters()[name]
    parameter._ignored = True
```

And then you don't need to leak your bookeeping of `ignored_parameter_list` anywhere else (You could even get rid of it, potentially).

And in dynamo, you would just do:

`p.requires_grad and not p._ignored`:

Comment (wconstab) [reply torch/_dynamo/eval_frame.py:246]: Thanks @voznesenskym, i think that's a great idea

Comment (wconstab) [reply torch/_dynamo/eval_frame.py:246]: cc @aazzolini @mrshenli any issues with this approach?

Comment (mrshenli) [reply torch/_dynamo/eval_frame.py:246]: SGTM. This is also what we are proposing for the new annotation-based API: https://fb.quip.com/bpvPA6f2dtrA

The only thing is that, we might want this at parameter level (instead of module-level) for DDP to have parity.

Comment (wconstab) [reply torch/_dynamo/eval_frame.py:246]: > The only thing is that, we might want this at parameter level (instead of module-level) for DDP to have parity.

I thought it already was at the parameter level?  See my latest code, i think it's what you want.

But now i'm confused- in DDPOptimizer I simply ignore all buffers, since i thought the implication was they never require grad, and thus wouldn't be allreduced by DDP.  If some buffers get allreduced by DDP, then i'd want to follow this up with another PR that tests buffers and gets that behavior right.

For now i've marked both params/buffers that are in the `parameters_and_buffers_to_ignore` list with the same marker on the DDP side, since that seems consistent with the convention there.

Comment (ngimel) [reply torch/_dynamo/eval_frame.py]: @alband says we can't rely on parameter id's being stable, that mostly works, but there are a few edge cases where it doesn't. In particular, reparametrization cannot always preserve original parameter id.

File: torch/nn/parallel/distributed.py [hunk 7]
```diff
@@ -1797,6 +1797,12 @@ def _set_params_and_buffers_to_ignore_for_model(
         # during synchronization. It will be removed when the API is finalized
         # as part of addressing https://github.com/pytorch/pytorch/issues/43690.
         module._ddp_params_and_buffers_to_ignore = params_and_buffers_to_ignore
+        for name, param in module.named_parameters():
+            if name in params_and_buffers_to_ignore:
+                param._ddp_ignored = True
+        for name, buffer in module.named_buffers():
```
Comment (wconstab) [inline torch/nn/parallel/distributed.py:1803]: @mrshenli i thought buffers by definition do not require grad, and therefore ddp ignores them by default?

if not i should update the logic in DDPOptimizer accordingly.

File: torch/_dynamo/eval_frame.py (hunk 6 above)
Comment (wconstab) [reply torch/_dynamo/eval_frame.py]: what does 'reparametrization' mean exactly?

i'm thinking it might be worth sticking with the current approach (marking params) as it is simple, and the consequences of getting it wrong are relatively minor (graph-breaks wouldn't exactly match ddp's buckets, so perf would degrade anywhere from a little bit to matching dynamo+ddp without graphbreaks.

but if there is another scheme that is not too complex i'd be open to it

Comment (ngimel) [reply torch/_dynamo/eval_frame.py]: Reparametrization is when you register a rule to recompute parameter every time before it's used https://pytorch.org/tutorials/intermediate/parametrizations.html

File: torch/nn/parallel/distributed.py (hunk 7 above)
Comment (aazzolini) [reply torch/nn/parallel/distributed.py:1803]: Buffers are by default broadcast right before the forward pass if  broadcast_buffers=True
is passed to DDP constructor. This is true by default.
But then, if the buffer appear on the igore_parameters field, it's not part of the broadcast.
I think buffers shouldn't count for the purposes of splitting the model since we're not syncing them after the backward pass.

File: torch/_dynamo/optimizations/distributed.py [hunk 8]
```diff
@@ -88,17 +96,22 @@ def compile_fn(self, gm: fx.GraphModule, example_inputs: List[torch.Tensor]):
             if node.op == "call_module":
                 target = gm.get_submodule(node.target)
                 for name, p in target.named_parameters():
-                    if p.requires_grad:
+                    param = target.get_parameter(name)
+                    if p.requires_grad and not self._ignore_parameter(param):
                         buckets[0].size += p.storage().nbytes()
-                        # TODO correct FQ name?
-                        buckets[0].params.append(f"{node}_{name}")
+                        buckets[0].params.append(f"{node.target}_{name}")
```
Comment (aazzolini) [inline torch/_dynamo/optimizations/distributed.py:102]: do we still need this?

Comment (wconstab) [reply torch/_dynamo/optimizations/distributed.py:102]: it is just useful for visualization purposes.  see the debug output on the next PR in this stack.  the buckets table is printed using this string.

File: torch/_dynamo/eval_frame.py (hunk 6 above)
Comment (wconstab) [reply torch/_dynamo/eval_frame.py]: ok. i think i want to propose that we just land this as is.
 - i'm not sure any of the users of DDP's ignored parameters flag today are also using parametrizations
 - it wouldn't be catastrophic if ignored_parameters was not honored in ddp optimizer. (that is the defacto today)
 - we could potentially revisit this later

Also, I'm curious-does dynamo+AOT handle parametrization on its own currently?

**Review by aazzolini — APPROVED** [review]
let's land as is but can you please add a _lot_ of comments on the code explaining where the logic breaks down and how we could solve it etc?
